### ISSUE TITLE
The accuracy is always zero while the loss is decreasing(ImageNet)

### ISSUE THREAD

**Issue by PkuRainBow** [deepinsight/insightface#117]
Great work, and I try to use the proposed loss to train imagenet 

> loss-type=4: Ours(InsightFace)

However,  it seems that the loss is decreasing but the accuracy is always zero?

Could anyone tell me the possible reasons?

**Comment by nttstar** [comment]
Because we decrease the value on the target logits with correct label. You have to wait for more iterations.A small margin is recommended if you want to train normal classification tasks such as imagenet.

**Comment by PkuRainBow** [comment]
@nttstar  I use a smaller margin 0.35 and train for a night, but the accuracy keeps zero ........ Very strange, I just remove some operations and the accuracy will become reasonable.

```
elif args.loss_type==4:
    s = args.margin_s
    m = args.margin_m
    assert s>0.0
    assert m>=0.0
    assert m<(math.pi/2)
    _weight = mx.symbol.Variable("fc7_weight", shape=(args.num_classes, args.emb_size), lr_mult=1.0)
    _weight = mx.symbol.L2Normalization(_weight, mode='instance')
    nembedding = mx.symbol.L2Normalization(embedding, mode='instance', name='fc1n')*s
    fc7 = mx.sym.FullyConnected(data=nembedding, weight = _weight, no_bias = True, num_hidden=args.num_classes, name='fc7')

 // the below operation will cause the accuracy always zero 
    zy = mx.sym.pick(fc7, gt_label, axis=1)
    cos_t = zy/s
    if args.margin_verbose>0:
      margin_symbols.append(mx.symbol.mean(cos_t))
    if args.output_c2c==0:
      cos_m = math.cos(m)
      sin_m = math.sin(m)
      mm = math.sin(math.pi-m)*m
      #threshold = 0.0
      threshold = math.cos(math.pi-m)
      if args.easy_margin:
        cond = mx.symbol.Activation(data=cos_t, act_type='relu')
      else:
        cond_v = cos_t - threshold
        cond = mx.symbol.Activation(data=cond_v, act_type='relu')
      body = cos_t*cos_t
      body = 1.0-body
      sin_t = mx.sym.sqrt(body)
      new_zy = cos_t*cos_m
      b = sin_t*sin_m
      new_zy = new_zy - b
      new_zy = new_zy*s
      if args.easy_margin:
        zy_keep = zy
      else:
        zy_keep = zy - s*mm
      new_zy = mx.sym.where(cond, new_zy, zy_keep)

```

**Comment by PkuRainBow** [comment]
I just have tried to set the margin-m as 0.1 and 0.3, but all of them failed.
Tuning the margin-s also failed.......      Very interesting


```
parser.add_argument('--margin-m', type=float, default=0.3, help='') # 0.5
parser.add_argument('--margin-s', type=float, default=20.0, help='') #64
parser.add_argument('--margin-a', type=float, default=1.0, help='')
parser.add_argument('--margin-b', type=float, default=0.0, help='')
parser.add_argument('--easy-margin', type=int, default=0, help='')
```

**Comment by bruinxiong** [comment]
The same issue, is there any solution for zero accuracy ?

**Comment by nttstar** [comment]
what's your batch-size? @PkuRainBow @bruinxiong

**Comment by nttstar** [comment]
Try give a small margin like 0.1 and use large batch-size(>128, 512 is recommended). Then wait for thousands of iterations.

**Comment by bruinxiong** [comment]
@nttstar Thank you for your reply. I set --per-batch-size = 64 so batch-size = 256 due to I have 4 GPUs. I cannot set more larger --per-batch-size, e.g. 128, due to the issue of out-of-memory, although my gpu is P100 with 16 GB memory.

**Comment by PkuRainBow** [comment]
I have tried m=0, 0.1, 0.2, 0.3, 0.4, 0.5， only the accuracy for m=0 seems normal. All the other choices failed and the accuracy is always zero even after 100 epochs(total is 120 epochs). Besides, it seems that the accuracy is very pool even when m=0.

**Comment by nttstar** [comment]
I have not tested the margin based algorithms on classification tasks. I will try it when I have time.

**Comment by ccl-private** [comment]
@nttstar any progress ？4 years ago. ~
